### PR TITLE
Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>

### PR DESCRIPTION
**Pull request by pbelevich**
This PR prepares `at::Generator` for pybind11's `type_caster<at::Generator>` which is required to implement custom RNG in python. The following changes are done:
1. `at::Generator` was moved to `c10::GeneratorImpl` (similar to `c10::TensorImpl`)
2. `at::Generator` was recreated as a holder of `std::shared_ptr<c10::GeneratorImpl>` (similar to `at::Tensor` that holds `c10::intrusive_ptr<c10::TensorImpl>`)
3. Most of `at::Generator*` usages were replaced with `at::Generator`

TBD: replacing `Generator generator = nullptr` with `{}` requires JIT changes(adding Generator to IValue?)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35167 Move normal() to DistributionTemplates
* #34774 pybind11 type_caster for at::Generator and custom RNG python test
* **#34468 Replace Generator* with Generator that holds std::shared_ptr<GeneratorImpl>**

Differential Revision: [D20549420](https://our.internmc.facebook.com/intern/diff/D20549420)

### CODE REVIEW

File: aten/src/ATen/CPUGenerator.cpp
Comment (ezyang) [inline aten/src/ATen/CPUGenerator.cpp]: Not really sure if you should be returning `Generator` here or `const Generator&` here. Latter might be better as this generator always stays live and there's no reason to force a refcount bump here.

File: aten/src/ATen/CPUGenerator.h
Comment (ezyang) [inline aten/src/ATen/CPUGenerator.h]: don't have to do it in this PR, but CPUGenerator and friends should probably also get the Impl suffix too

File: aten/src/ATen/Context.h
Comment (ezyang) [inline aten/src/ATen/Context.h]: Also doesn't have to be this PR, but we should consider hiding the mutex behind a method instead of exposing all the field guts to everyone.

File: aten/src/ATen/Utils.h
Comment (ezyang) [inline aten/src/ATen/Utils.h]: These can take the generator by reference, the function doesn't take ownership

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: A much safer variant of this overload is `Generator(nullptr_t gen_impl)` which will only permit `nullptr` exactly.

Comment (ezyang) [reply aten/src/ATen/core/Generator.h]: (If I read the TODO correctly, that is!)

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: You shouldn't need to define these explicitly; the default implementations should do the right thing.

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: This is rarely what you actually want, right, because the abstract `GeneratorImpl` class generally doesn't have enough information for you to do anything useful with it. Instead, I'd suggest axing this method, and make the typed cast to CPUGenerator/CUDAGenerator template function a method on Generator itself.

File: aten/src/ATen/cuda/CUDAGenerator.cpp
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGenerator.cpp]: With my C++ API hat on, we should endeavor to make the API for allocating a CUDA generator look as much as possible as the Python API. (So, not `make_generator`, in general! `make_generator` still a useful helper function for internal implementation reasons though, I guess.)

File: aten/src/ATen/cuda/CUDAGenerator.cpp
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGenerator.cpp]: This is kind of goofy. Better to just check generator once and then use the resulting handle to do the mutations you need, right?

File: aten/src/ATen/test/cpu_generator_test.cpp
Comment (ezyang) [inline aten/src/ATen/test/cpu_generator_test.cpp]: Why is the `Generator()` constructor call needed here?

File: aten/src/ATen/test/rng_test.h
Comment (ezyang) [inline aten/src/ATen/test/rng_test.h]: Music to my ears!!

File: torch/csrc/Generator.cpp
Comment (ezyang) [inline torch/csrc/Generator.cpp]: Whoops! You're leaking the generators here. Because the PyObject now directly contains a C++ type, you must placement delete it on deallocation.

File: torch/csrc/Generator.h
Comment (ezyang) [inline torch/csrc/Generator.h]: nit: just delete this

**Review by ezyang — CHANGES_REQUESTED** [review]
This is great work, a giant step in the right direction. Unfortunately, there's a memory leak which needs to be fixed before this can go in.

Some broader things to keep in mind (they don't have to be done in this PR):
1. We should consider taking `const Generator&` instead of `Generator` as it avoids a refcount bump. (cc @zdevito for any extra opinions)
2. We should think about how to make it easy for user code to do defaulting behavior correctly for generators. This is kind of entwined with Tensor, which has a default constructor, but for which in our user API we encourage people to accept `optional<Tensor>` in cases when the Tensor may not be specified. In the case of generator, the right call might be to NOT use optional, but we need to make it easy for people to avoid accidentally doing a null pointer dereference. I don't think this should block this PR either, as null pointer was always a problem for the old API too.

File: torch/csrc/Generator.h
Comment (pbelevich) [reply torch/csrc/Generator.h]: addressed

File: torch/csrc/Generator.cpp
Comment (pbelevich) [reply torch/csrc/Generator.cpp]: addressed

File: aten/src/ATen/test/cpu_generator_test.cpp
Comment (pbelevich) [reply aten/src/ATen/test/cpu_generator_test.cpp]: Because `clone()` is a method of `at::GeneratorImpl` and returns `std::shared_ptr<at::GeneratorImpl>` and c-tor `at::Generator(std::shared_ptr<at::GeneratorImpl>)` is explicit

File: aten/src/ATen/cuda/CUDAGenerator.cpp
Comment (pbelevich) [reply aten/src/ATen/cuda/CUDAGenerator.cpp]: addressed

File: aten/src/ATen/test/cpu_generator_test.cpp
Comment (ezyang) [reply aten/src/ATen/test/cpu_generator_test.cpp]: Maybe there should be a `Generator::clone()` for this case?

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: removed

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: addressed

File: aten/src/ATen/Utils.h
Comment (pbelevich) [reply aten/src/ATen/Utils.h]: addressed

File: aten/src/ATen/CPUGenerator.cpp
Comment (pbelevich) [reply aten/src/ATen/CPUGenerator.cpp]: addressed

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: addressed

File: aten/src/ATen/cuda/CUDAGenerator.cpp
Comment (pbelevich) [reply aten/src/ATen/cuda/CUDAGenerator.cpp]: possible solution is adding `Generator(Device device)`, please see the next PR

File: aten/src/ATen/test/cpu_generator_test.cpp
Comment (pbelevich) [reply aten/src/ATen/test/cpu_generator_test.cpp]: added

File: aten/src/ATen/CPUGenerator.h
Comment (pbelevich) [reply aten/src/ATen/CPUGenerator.h]: #34984

File: aten/src/ATen/Context.h
Comment (pbelevich) [reply aten/src/ATen/Context.h]: #34987

File: aten/src/ATen/cuda/CUDAGenerator.cpp
Comment (ezyang) [reply aten/src/ATen/cuda/CUDAGenerator.cpp]: Constructor with device is probably not the way we should do it :) We'll talk about it on the other PR.

File: torch/csrc/Generator.cpp
Comment (ezyang) [reply torch/csrc/Generator.cpp]: First, sorry, I misspoke in my comment; it's not placement delete you need to do (placement delete is not a thing), you need to just explicitly call the destructor. Assignment with the default constructor isn't sufficient, as a default constructed Generator may still have a nontrivial destructor.

```
self->cdata->~Generator();
```

File: aten/src/ATen/Context.h
Comment (ezyang) [inline aten/src/ATen/Context.h]: Also doesn't have to be this PR, but we should consider adding methods like `set_current_seed` to Generator directly, so dot-syntax works on them (making Generator consistent with Tensor)

File: aten/src/ATen/Utils.h
Comment (ezyang) [inline aten/src/ATen/Utils.h]: When can this be nullptr?

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: It's pretty important for us to get rid of this TODO soon, as it's pretty confusing that:

```
std::shared_ptr<c10::GeneratorImpl> impl = nullptr;
Generator g(impl); // errors

Generator g(nullptr); // this is OK?!
```

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: Can't this just be `this->impl_ == rhs->impl_`? Because if they're both nullptr they'll compare equal.

File: aten/src/ATen/core/Generator.h
Comment (ezyang) [inline aten/src/ATen/core/Generator.h]: nit: `static_cast<bool>(impl_)` for extra safety. (Or just let implicit conversion on pointer take care of it.)

**Review by ezyang — APPROVED** [review]
LGTM.

**Comment by ezyang** [comment]
Please take care of replacing the `= {}` with explicit destructor call before landing

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: addressed

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: addressed

File: aten/src/ATen/Context.h
Comment (pbelevich) [reply aten/src/ATen/Context.h]: will do in #34987

File: aten/src/ATen/core/Generator.h
Comment (pbelevich) [reply aten/src/ATen/core/Generator.h]: Totally agree, #35067

File: aten/src/ATen/Utils.h
Comment (pbelevich) [reply aten/src/ATen/Utils.h]: I rewrote this part, please take a look

**Pull request closed without merging**
